### PR TITLE
gh-102936: typing: document performance pitfalls of protocols decorated with `@runtime_checkable`

### PR DESCRIPTION
**Pull request by AlexWaygood**
* Issue: gh-102936

### CODE REVIEW

**Review by gvanrossum — APPROVED** [review]
Makes sense.

File: Doc/library/typing.rst [hunk 1]
```diff
@@ -1584,16 +1584,31 @@ These are not used in annotations. They are building blocks for creating generic
 
       assert isinstance(open('/some/file'), Closable)
 
+      @runtime_checkable
```
Comment (hauntsaninja) [inline Doc/library/typing.rst:1587]: I think many users aren't too familiar with `OSError`'s constructor / we can't use kwargs and it's unclear what the second argument is. Maybe we could use `int` and `HasNumerator` instead?

(I know it's not writeable, but `OSError`'s errno also is maybe not writeable? Writing to errno doesn't affect repr)

File: Doc/library/typing.rst [hunk 2]
```diff
@@ -1584,16 +1584,31 @@ These are not used in annotations. They are building blocks for creating generic
...
         check against :data:`Callable`.  However, the
         ``ssl.SSLObject.__init__`` method exists only to raise a
         :exc:`TypeError` with a more informative message, therefore making
         it impossible to call (instantiate) :class:`ssl.SSLObject`.
 
+   .. note::
+
+        An :func:`isinstance` check against a runtime-checkable protocol can be
+        surprisingly slow compared to an ``isinstance()`` check against
+        a non-protocol class. Consider using alternative idioms such as
```
Comment (hauntsaninja) [inline Doc/library/typing.rst:1609]: Great as is, but if we want to sow further doubt:
```suggestion
        a non-protocol class (and very difficult to spot on a profile). Consider using alternative idioms such as
```

**Review by hauntsaninja — COMMENTED** [review]
Thanks for doing this! Had one suggestion for a different example of attribute

File: Doc/library/typing.rst (hunk 1 above)
Comment (AlexWaygood) [reply Doc/library/typing.rst:1587]: Yeah, I was really struggling to think of a good example for this! I'll ponder some more and see if I can come up with something better...

File: Doc/library/typing.rst (hunk 2 above)
Comment (AlexWaygood) [reply Doc/library/typing.rst:1609]: Hmm, I'm inclined to leave this one as is — not sure it's worth the added verbosity, and I want to get on as quickly as possible to "here's what to do instead"

File: Doc/library/typing.rst (hunk 1 above)
Comment (hauntsaninja) [reply Doc/library/typing.rst:1587]: Perhaps unsurprisingly, not too many good cases of writeable attributes in stdlib. Maybe Thread.name? It's explicitly documented as something you can write to.

Comment (AlexWaygood) [reply Doc/library/typing.rst:1587]: That works! I'll make the change tomorrow (it's late here)

**Review by furkanonder — APPROVED** [review]

**Review by hauntsaninja — APPROVED** [review]

**Comment by AlexWaygood** [comment]
Thanks all for the reviews!

**Pull request merged**